plugins/SpectrumAnalyzer/SaProcessor.h

```diff
@@ -72,15 +78,27 @@ class SaProcessor
 	float getAmpRangeMin(bool linear = false) const;
 	float getAmpRangeMax() const;
 
-	// data access lock must be acquired by any friendly class that touches
-	// the results, mainly to prevent unexpected mid-way reallocation
+	// Reallocation lock prevents the processor from changing size of its buffers.
+	// It is used to keep consistent bin-to-frequency mapping while drawing the
+	// spectrum and to make sure reading side does not find itself out of bounds.
+	// The processor is meanwhile free to work on another block.
+	QMutex m_reallocationAccess;
+	// Data access lock prevents the processor from changing both size and content
+	// of its buffers. It is used when writing to a result buffer, or when a friendly
+	// class reads them and needs guaranteed data consistency.
+	// It causes FFT analysis to be paused, so this lock should be used sparingly.
+	// If using both locks at the same time, reallocation lock MUST be acquired first.
 	QMutex m_dataAccess;
```

Comment:
Until now, I did not yet check whether the mutexes where always used when vulnerable data was accessed, because. I'm not sure which data vulnerable data must be protected by which mutex. Maybe this should be noted in this (and the above) comment. Also, it might help to remove the `friend` definitions from `SaProcessor`.

Reply:
I thought that is kind of self-evident from the description: reallocation lock is used to signal that buffer size must not change, so it applies to all buffers that have any size, and to variables that describe that size.
Data access lock is used to guarantee that two threads don't write to a buffer at the same time, producing unpredictable results. Originally it was also used to prevent the FFT thread from changing results buffers while GUI thread was reading them, but that lead to poor performance.

That means there may be some buffers that are left unprotected by the data lock, mainly in the "view" classes: I don't remember exactly (as I wrote that all some time ago), but I tried to get rid of as much locking as possible to prevent GUI from stalling. Any data corruption resulting from results being updated during drawing should be small and rare enough that the one slightly damaged frame once in a while won't be noticeable.

The reallocation lock on the other hand _should_ cover everything, since if a buffer changes its size in the middle of a for loop, it would probably result in a noticeable mob of angry users complaining about segfaults.

Reply:
Thanks, I'll have to re-check that in the next days.

Reply:
Thanks, I'll have to re-check this in the next days.